### PR TITLE
[react-map-gl]Add type definitions of source and layer components

### PR DESCRIPTION
**Pull request by williamlabdev**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://uber.github.io/react-map-gl/#/Documentation/api-reference/source>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/react-map-gl/index.d.ts [hunk 1]
```diff
@@ -423,3 +424,47 @@ export interface SVGOverlayProps extends BaseControlProps {
 }
 
 export class SVGOverlay extends BaseControl<SVGOverlayProps, Element> {}
+
+export interface SourceProps {
```
Comment (fnberta) [inline types/react-map-gl/index.d.ts]: Where did you get all these props from? In the documentation and the source code, I only see `id`, `type`, and `children`. Or am I missing something?

**Review by fnberta — CHANGES_REQUESTED** [review]
Thank you for this PR. I have one question regarding the Props of the `Source` component, see inline.

File: types/react-map-gl/index.d.ts (hunk 1 above)
Comment (chiuhow) [reply types/react-map-gl/index.d.ts]: In the source code of react-map-gl ,sourceOptions is provdied implictly and extra properties will be added through the loop.
https://github.com/uber/react-map-gl/blob/c71b0d4b9da65bdeaa4a077d85b4658654b68100/src/components/source.js#L96
I add these extra properties and set them as optional and people could pass extra properties which are defined by mapbox-gl with type checking.

**Review by fnberta — APPROVED** [review]
Sorry for the delay, approving!

File: types/react-map-gl/index.d.ts (hunk 1 above)
Comment (fnberta) [reply types/react-map-gl/index.d.ts]: Ok, cool

**Pull request merged**